cooked-validators/src/Cooked/MockChain/Monad/Staged.hs

```diff
@@ -57,196 +56,87 @@ type InterpMockChain = MockChainT (WriterT TraceDescr [])
---  sure the resulting monad will be an instance of 'MonadFail'.
-data MockChainOp a where
-  ValidateTxSkel :: TxSkel -> MockChainOp Pl.TxId
-  TxOutByRef :: Pl.TxOutRef -> MockChainOp (Maybe Pl.TxOut)
-  GetCurrentSlot :: MockChainOp Pl.Slot
-  AwaitSlot :: Pl.Slot -> MockChainOp Pl.Slot
-  GetCurrentTime :: MockChainOp Pl.POSIXTime
-  AwaitTime :: Pl.POSIXTime -> MockChainOp Pl.POSIXTime
+data MockChainBuiltin a where
+  ValidateTxSkel :: TxSkel -> MockChainBuiltin Pl.TxId
+  TxOutByRef :: Pl.TxOutRef -> MockChainBuiltin (Maybe Pl.TxOut)
+  GetCurrentSlot :: MockChainBuiltin Pl.Slot
+  AwaitSlot :: Pl.Slot -> MockChainBuiltin Pl.Slot
+  GetCurrentTime :: MockChainBuiltin Pl.POSIXTime
+  AwaitTime :: Pl.POSIXTime -> MockChainBuiltin Pl.POSIXTime
   UtxosSuchThat ::
     (Pl.FromData a) =>
     Pl.Address ->
     (Maybe a -> Pl.Value -> Bool) ->
-    MockChainOp [(SpendableOut, Maybe a)]
-  OwnPubKey :: MockChainOp Pl.PubKeyHash
-  --
-  SigningWith :: NE.NonEmpty Wallet -> StagedMockChain a -> MockChainOp a
-  AskSigners :: MockChainOp (NE.NonEmpty Wallet)
-  GetSlotConfig :: MockChainOp Pl.SlotConfig
-  --
-  Modify :: Modality TxSkel -> StagedMockChain a -> MockChainOp a
-  --
-  Fail :: String -> MockChainOp a
-  --
-  MPlus :: StagedMockChain a -> StagedMockChain a -> MockChainOp a
-  MZero :: MockChainOp a
+    MockChainBuiltin [(SpendableOut, Maybe a)]
+  OwnPubKey :: MockChainBuiltin Pl.PubKeyHash
+  -- the following are only available in MonadMockChain, not MonadBlockChain:
+  SigningWith :: NE.NonEmpty Wallet -> StagedMockChain a -> MockChainBuiltin a
+  AskSigners :: MockChainBuiltin (NE.NonEmpty Wallet)
+  GetSlotConfig :: MockChainBuiltin Pl.SlotConfig
 
-data Staged (op :: * -> *) :: * -> * where
-  Return :: a -> Staged op a
-  Instr :: op a -> (a -> Staged op b) -> Staged op b
+type MockChainOp = LtlOp Attack MockChainBuiltin
 
 type StagedMockChain = Staged MockChainOp
 
-singleton :: MockChainOp a -> StagedMockChain a
-singleton op = Instr op Return
-
-instance Functor StagedMockChain where
-  fmap f (Return x) = Return $ f x
-  fmap f (Instr op cont) = Instr op (fmap f . cont)
+-- * 'InterpLtl' instance
 
-instance Applicative StagedMockChain where
-  pure = Return
-  (<*>) = ap
+instance {-# OVERLAPS #-} Semigroup Attack where
+  f <> g = maybe Nothing f . g
 
-instance Monad StagedMockChain where
-  (Return x) >>= f = f x
-  (Instr i m) >>= f = Instr i (m >=> f)
+instance {-# OVERLAPS #-} Monoid Attack where
+  mempty = Just
 
-instance MonadFail StagedMockChain where
-  fail = singleton . Fail
+instance MonadPlus m => MonadPlus (MockChainT m) where
+  mzero = lift mzero
+  mplus = combineMockChainT mplus
 
-instance Alternative StagedMockChain where
-  empty = singleton MZero
-  a <|> b = singleton $ MPlus a b
+instance InterpLtl Attack MockChainBuiltin InterpMockChain where
+  interpBuiltin (ValidateTxSkel skel) =
+    get
+      >>= msum
+        . map (\(now, later) -> put later >> maybe mzero validateTxSkel (now skel))
```

Comment:
```suggestion
        . map (\(now, later) -> maybe mzero validateTxSkel (now skel) <* put later)
```

Giant nitpick; but it feels a little more natural to me to see `put later` *after* the `validateTxSkel` took place.